File: public/java/047-connected-components.java
/*
 * LeetCode Problem #323: Number of Connected Components in an Undirected Graph
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/
 */
import java.util.*;

public class Solution {

    // APPROACH 1: DFS | O(V+E) time | O(V+E) space
    // EXPLAIN: DFS from each unvisited node; each invocation discovers one connected component.
    public int countComponents(int n, int[][] edges) {
        List<List<Integer>> adj = buildAdj(n, edges);
        boolean[] visited = new boolean[n];
        int count = 0;
        for (int i = 0; i < n; i++) if (!visited[i]) { dfs(i, adj, visited); count++; }
        return count;
    }
    private void dfs(int u, List<List<Integer>> adj, boolean[] visited) {
        visited[u] = true;
        for (int v : adj.get(u)) if (!visited[v]) dfs(v, adj, visited);
    }

    // APPROACH 2: BFS | O(V+E) time | O(V+E) space
    // EXPLAIN: BFS from each unvisited node; count how many BFS starts are needed.
    public int countComponentsBFS(int n, int[][] edges) {
        List<List<Integer>> adj = buildAdj(n, edges);
        boolean[] visited = new boolean[n];
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                count++;
                Queue<Integer> q = new LinkedList<>();
                q.offer(i); visited[i] = true;
                while (!q.isEmpty()) {
                    int u = q.poll();
                    for (int v : adj.get(u)) if (!visited[v]) { visited[v] = true; q.offer(v); }
                }
            }
        }
        return count;
    }

    // APPROACH 3: Union-Find | O(E·α(V)) time | O(V) space
    // EXPLAIN: Start with n components; each successful union decrements the count.
    public int countComponentsUF(int n, int[][] edges) {
        int[] parent = new int[n], rank = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;
        int count = n;
        for (int[] e : edges) {
            int ra = find(parent, e[0]), rb = find(parent, e[1]);
            if (ra != rb) {
                if (rank[ra] < rank[rb]) { int t = ra; ra = rb; rb = t; }
                parent[rb] = ra;
                if (rank[ra] == rank[rb]) rank[ra]++;
                count--;
            }
        }
        return count;
    }
    private int find(int[] p, int x) { return p[x]==x?x:(p[x]=find(p,p[x])); }

    private List<List<Integer>> buildAdj(int n, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int[] e : edges) { adj.get(e[0]).add(e[1]); adj.get(e[1]).add(e[0]); }
        return adj;
    }
}

// Made with Bob

File: public/python/047-connected-components.py
# LeetCode Problem #323: Number of Connected Components in an Undirected Graph
# Difficulty: Medium
# Link: https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/

from typing import List
from collections import deque


# ─────────────────────────────────────────────
# APPROACH 1: DFS | O(V+E) time | O(V+E) space
# EXPLAIN: DFS from each unvisited node; each DFS call discovers one connected component.
# WHEN: Simple recursive approach; standard first answer.

def countComponents_dfs(n: int, edges: List[List[int]]) -> int:
    adj = [[] for _ in range(n)]
    for u, v in edges:
        adj[u].append(v)
        adj[v].append(u)
    visited = [False] * n

    def dfs(node):
        visited[node] = True
        for nb in adj[node]:
            if not visited[nb]:
                dfs(nb)

    count = 0
    for i in range(n):
        if not visited[i]:
            dfs(i)
            count += 1
    return count


# ─────────────────────────────────────────────
# APPROACH 2: BFS | O(V+E) time | O(V+E) space
# EXPLAIN: BFS from each unvisited node; count how many BFS starts are needed.
# WHEN: Iterative; avoids recursion depth issues.

def countComponents_bfs(n: int, edges: List[List[int]]) -> int:
    adj = [[] for _ in range(n)]
    for u, v in edges:
        adj[u].append(v)
        adj[v].append(u)
    visited = [False] * n
    count = 0
    for i in range(n):
        if not visited[i]:
            count += 1
            q = deque([i])
            visited[i] = True
            while q:
                node = q.popleft()
                for nb in adj[node]:
                    if not visited[nb]:
                        visited[nb] = True
                        q.append(nb)
    return count


# ─────────────────────────────────────────────
# APPROACH 3: Union-Find | O(E·α(V)) time | O(V) space
# EXPLAIN: Start with n components; each union of two different components decrements the count.
# WHEN: Optimal; naturally models dynamic connectivity.

def countComponents_union_find(n: int, edges: List[List[int]]) -> int:
    parent = list(range(n))
    rank = [0] * n

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra == rb:
            return False
        if rank[ra] < rank[rb]:
            ra, rb = rb, ra
        parent[rb] = ra
        if rank[ra] == rank[rb]:
            rank[ra] += 1
        return True

    count = n
    for u, v in edges:
        if union(u, v):
            count -= 1
    return count


# Made with Bob
